File: test/test_imagedistancematrix.js
/*  LICENSE
    
    _This file is Copyright 2018 by the Image Processing and Analysis Group (BioImage Suite Team). Dept. of Radiology & Biomedical Imaging, Yale School of Medicine._
    
    BioImage Suite Web is licensed under the Apache License, Version 2.0 (the "License");
    
    - you may not use this software except in compliance with the License.
    - You may obtain a copy of the License at [http://www.apache.org/licenses/LICENSE-2.0](http://www.apache.org/licenses/LICENSE-2.0)
    
    __Unless required by applicable law or agreed to in writing, software
    distributed under the License is distributed on an "AS IS" BASIS,
    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
    See the License for the specific language governing permissions and
    limitations under the License.__
    
    ENDLICENSE */

/* jshint node:true */
/*global describe, it, before */
"use strict";

require('../config/bisweb_pathconfig.js');
const assert = require("assert");
const BisWebImage=require('bisweb_image');
const BisWebMatrix=require('bisweb_matrix');
const path=require('path');
const libbiswasm=require('libbiswasm_wrapper');
const tempfs = require('temp').track();

const tmpDirPath=tempfs.mkdirSync('test_image');
const tmpFname1=path.resolve(tmpDirPath,'matrix1.binmatr');
const tmpFname2=path.resolve(tmpDirPath,'matrix2.binmatr');

let gold = [ new BisWebMatrix(),
             new BisWebMatrix(),
             new BisWebMatrix()];
let matnames = [ 'sparse.matr',
                 'radius.matr' ,
                 'radius2.matr' ];

let img=null,obj=null,indexmap=null;

describe('Testing imageDistanceMatrix stuff\n', function() {

    this.timeout(50000);
    
    before(function(done){

        for (let i=0;i<=2;i++) 
            matnames[i]=path.resolve(__dirname, 'testdata/distancematrix/'+matnames[i]);


        let p = [ libbiswasm.initialize() ,
                  gold[0].load(matnames[0]),
                  gold[1].load(matnames[1]),
                  gold[2].load(matnames[2])
                ];

        img=new BisWebImage();
        img.createImage( {
            type : 'float',
            dimensions : [ 5,5,1],
            spacing : [ 2.0,2.0,2.0 ]
        });
        console.log('....',img.getDescription());

        let dat=img.getImageData();
        for (let i=0;i<25;i++)
            dat[i]=i;

        obj=new BisWebImage();
        obj.createImage( {
            type : 'short',
            dimensions : [ 5,5,1],
            spacing : [ 2.0,2.0,2.0 ]
        });

        indexmap=new BisWebImage();
        indexmap.cloneImage(obj);

        console.log('....',obj.getDescription());
        let odat=obj.getImageData();
        let idat=indexmap.getImageData();
        let index=1;
        for (let j=0;j<5;j++) {
            for (let i=0;i<5;i++) {
                if (i>=1 && i<4 && j>=1 && j<4) {
                    odat[i+j*5]=1;
                    idat[i+j*5]=index;
                    index=index+1;
                } else {
                    odat[i+j*5]=0;
                }
            }
        }
        console.log('____',idat);
        console.log('____',odat);

        
        Promise.all(p).then( () => {

            console.log('\n\n---------------------------------------------------\n\n');
            done();
        });
    });

   it ('test distmatrix1',function() {

       return new Promise( (resolve) => {
           let out=libbiswasm.computeImageIndexMapWASM(obj,true);
           let result=out.compareWithOther(indexmap);
           console.log('....',result);
           assert.equal(true,result.testresult);
           resolve();
       });
    });

    it ('test distmatrix2',function() {

        return new Promise( (resolve) => {

            let out2=libbiswasm.computeImageDistanceMatrixWASM(img,obj,
                                                      { "useradius" : true,
                                                        "radius" : 3.0,
                                                        "numthreads" : 1
                                                      },0);
            let result=out2.compareWithOther(gold[0]);
            console.log('....',result);
            assert.equal(true,result.testresult);
            resolve();
        });
        
    });

    it ('test distmatrix3',function() {
        return new Promise( (resolve) => {
            let out3=libbiswasm.computeImageDistanceMatrixWASM(img,obj,{ "useradius" : false,
                                                                         "sparsity" : 0.1,
                                                                         "numthreads" : 1
                                                                       },1);

            let result=out3.compareWithOther(gold[2]);
            console.log('....',result,' type=',out3.data.constructor.name,out3.datatype);
            assert.equal(true,result.testresult);
            resolve();
        });
    });

    it ('test load and save',function() {
        return new Promise( (resolve) => {
            gold[1].save(tmpFname1).then( () => {
                let sample =  new BisWebMatrix();
                //            gold[1].save('sample.binmatr').then( () => {
                sample.load(path.resolve(__dirname, 'testdata/distancematrix/sample.binmatr')).then( () => {
                    
                    let newmatr=new BisWebMatrix();
                    newmatr.load(tmpFname1).then( () => {
                        let result=newmatr.compareWithOther(gold[1]);
                        let result2=sample.compareWithOther(gold[1]);
                        console.log('....',result,result2);
                        let ok=false;
                        if (result.testresult && result2.testresult)
                            ok=true;
                        assert.equal(true,ok);
                        resolve();
                    });
                });
            });
        });
    });

    it ('test load and save double', function() {
        return new Promise( (resolve) => {
            let sample =  new BisWebMatrix();
            sample.load(path.resolve(__dirname, 'testdata/distancematrix/double.binmatr')).then( () => {
                sample.save(tmpFname2).then( () => {
                    let newmatr=new BisWebMatrix();
                    newmatr.load(tmpFname2).then( () => {
                        let result=newmatr.compareWithOther(sample);
                        let result2=sample.compareWithOther(gold[1]);
                        console.log('....',result,result2);
                        let ok=false;
                        if (result.testresult && result2.testresult)
                            ok=true;
                        assert.equal(true,ok);
                        resolve();
                    });
                });
            });
        });
    });

});

File: test/test_imagedistancematrix.py
# LICENSE
# 
# _This file is Copyright 2018 by the Image Processing and Analysis Group (BioImage Suite Team). Dept. of Radiology & Biomedical Imaging, Yale School of Medicine._
# 
# BioImage Suite Web is licensed under the Apache License, Version 2.0 (the "License");
# 
# - you may not use this software except in compliance with the License.
# - You may obtain a copy of the License at [http://www.apache.org/licenses/LICENSE-2.0](http://www.apache.org/licenses/LICENSE-2.0)
# 
# __Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.__
# 
# ENDLICENSE

import os
import sys
import numpy as np
import unittest
import tempfile

my_path=os.path.dirname(os.path.realpath(__file__));
sys.path.insert(0,os.path.abspath(my_path+'/../'));
sys.path.insert(0,os.path.abspath(my_path+'/../biswebpython/modules'));

from biswebpython.core.bis_objects import *
import biswebpython.core.bis_baseutils as bis_baseutils;

libbis=bis_baseutils.getDynamicLibraryWrapper();


gold = [ bisMatrix(), bisMatrix() ];
matnames = [ 'sparse.matr',   'radius.matr' ];

gold[1].load(my_path+'/../test/testdata/distancematrix/'+matnames[1]);
gold[0].load(my_path+'/../test/testdata/distancematrix/'+matnames[0]);


affine=np.eye(4,dtype=np.float32);
affine[0][0]=2.0;
affine[1][1]=2.0;
affine[2][2]=2.0;

idat=np.zeros([5,5],dtype=np.float32);
objdat=np.zeros([5,5],dtype=np.int16);
indexdat=np.zeros([5,5],dtype=np.int16);

print('Creating data');
index=1;
for j in range(0,5):
    for i in range(0,5):
        idat[i][j]=j*5+i;
        if (i>=1 and i<4 and j>=1 and j<4):
            objdat[i][j]=1;
            indexdat[i][j]=index;
            index=index+1;
            
print(idat);
print(objdat);
print(indexdat);

img=bisImage().create(idat,[2.0,2.0,2.0],affine);
objectmap=bisImage().create(objdat,[2.0,2.0,2.0],affine);
indexmap=bisImage().create(indexdat,[2.0,2.0,2.0],affine);

print(img.getDescription());
print(objectmap.getDescription());
print(indexmap.getDescription());

def computeNorm2(data1,data2):
    A=data1.flatten();
    B=data2.flatten();
    return np.linalg.norm(A-B);


class TestImageDistanceMatrix(unittest.TestCase):

    def test_indexmap(self):

        out=libbis.computeImageIndexMapWASM(objectmap,True);
        result=computeNorm2(out.data_array,indexmap.data_array);

        if result<1:
            testpass=True
        else:
            testpass=False;
        
        print('----------------------------------------------------------')
        print('__ indexmap computation diff=',result, 'pass=',testpass);
        print('----------------------------------------------------------')
        
        self.assertEqual(testpass,True);

    def test_radiusmatrix(self):

        print(img.getDescription());
        print(objectmap.getDescription());
        
        out2=libbis.computeImageDistanceMatrixWASM(img,objectmap,{ "useradius" : True,
                                                                   "radius" : 3.0,
                                                                   "numthreads" : 1 },0);
        result=computeNorm2(gold[0].data_array,out2);
        if result<1:
            testpass=True
        else:
            testpass=False;
        
        print('----------------------------------------------------------')
        print('__ radius distance matrix computation diff=',result, 'pass=',testpass);
        print('----------------------------------------------------------')
        
        self.assertEqual(testpass,True);

    def test_sparsematrix(self):

        print(img.getDescription());
        print(objectmap.getDescription());
        
        out2=libbis.computeImageDistanceMatrixWASM(img,objectmap,{ "useradius" : False,
                                                                   "sparsity" : 0.1,
                                                                   "numthreads" : 1 },0);
        result=computeNorm2(gold[1].data_array,out2);
        if result<1:
            testpass=True
        else:
            testpass=False;
        
        print('----------------------------------------------------------')
        print('__ sparse distance matrix computation diff=',result, 'pass=',testpass);
        print('----------------------------------------------------------')
        
        self.assertEqual(testpass,True);

    def test_sparsematrixmulti(self):

        print(img.getDescription());
        print(objectmap.getDescription());
        
        out2=libbis.computeImageDistanceMatrixWASM(img,objectmap,{ "useradius" : False,
                                                                   "sparsity" : 0.1,
                                                                   "numthreads" : 2 },0);
        result=computeNorm2(gold[1].data_array,out2);
        if result<1:
            testpass=True
        else:
            testpass=False;
        
        print('----------------------------------------------------------')
        print('__ multi sparse distance matrix computation diff=',result, 'pass=',testpass,' ',out2.dtype);
        print('----------------------------------------------------------')
        
        self.assertEqual(testpass,True);

    def test_loadsparse(self):

        fname=(my_path+'/../test/testdata/distancematrix/sample.binmatr');
        print(' ... loading',fname);
        binM = bisMatrix();
        binM.load(fname);

        with tempfile.TemporaryDirectory() as tempdname:
            out=tempdname+'/test.binmatr';
            binM.save(out);

            binM2 = bisMatrix();
            binM2.load(out);
            print('Loaded back from ',out);

        
            result=computeNorm2(gold[1].data_array,binM.data_array);
            result2=computeNorm2(gold[1].data_array,binM2.data_array);

            
            if result<1 and result2<1:
                testpass=True
            else:
                testpass=False;
        
        print('----------------------------------------------------------')
        print('__ bin load sparse matrix=',result, 'pass=',testpass);
        print('----------------------------------------------------------')
        
        self.assertEqual(testpass,True);

        self.assertEqual(True,True);
        

if __name__ == '__main__':
    TestResample().main()        
